Adds proxy for copy on write

ishanrai05 opened the pull request.

alexanderrichards commented on ganga/GangaCore/GPIDev/Base/Proxy.py:
```diff
@@ -1205,6 +1205,165 @@ def _getattribute(self, name):
+    attribute that was touched). This behavior is triggered by an appropriate
+    flag.
+    """
+    __slots__ = ["_obj", "__weakref__", "__slots__", "_is_copied",
```
You don't need to define `"__slots__"` within `__slots__`

alexanderrichards commented on ganga/GangaCore/GPIDev/Base/Proxy.py:
```diff
@@ -1205,6 +1205,165 @@ def _getattribute(self, name):
+
+    def __init__(self, obj, _partial_copy=True):
+        object.__setattr__(self, "_obj", obj)
+        object.__setattr__(self, "_enable_partial_copy", _partial_copy)
```
as long as the attribute names `_obj` and `_enable_partial_copy` are defined within `__slots__` (as they are) you should be able to assign these using the standard:
```python
self._obj = obj
self._enable_partial_copy = _partial_copy
```
rather than this unbounded (static) call to object base class's `__setattr__`.

alexanderrichards replied:
I see it's because you have overridden the `object.__setattr__` and you don't want it triggered on these instance variables

alexanderrichards commented on ganga/GangaCore/GPIDev/Base/Proxy.py:
```diff
@@ -1205,6 +1205,165 @@ def _getattribute(self, name):
+        theclass.__init__(ins, obj, *args, **kwargs)
+        return ins
+
+
```
My general feeling is that this is quite dense, especially with all the static calls to the object base class. I have thrown together a simple first approach. What are the deficiencies of the following?
```python
class COWProxy(object):
    __slots__ = ('_obj', '_partial', '_partial_dict', '_copied')

    def __init__(self, obj, partial=False):
        self._obj = obj
        self._partial = partial
        self._partial_dict = {}
        self._copied = False

    def __getattribute__(self, attr):
        if self._copied:
            return getattr(self._obj, attr)
        return self._partial_dict.setdefault(attr, COWProxy(getattr(self._obj, attr)))

    def __setattr__(self, attr, value):
        if attr in COWProxy.__slots__:
            return object.__setattr__(self, attr, value)
        if self._copied:
            return setattr(self._obj, attr, value)
        if not self._partial:
            self._obj = deepcopy(self._obj)
            self._copied = True
            return setattr(self._obj, attr, value)
        self._partial_dict[attr] = value
```

ishanrai05 committed:
Implements copy on write using flyingweight pattern

ishanrai05 committed:
Changed OrderedDict to dict

mesmith75 commented on ganga/GangaCore/GPIDev/Base/CoW.py:
```diff
@@ -0,0 +1,504 @@
+    # dict[<var_type>][__hash__]
+    _flyweight_cache = {}
+
+    def __init__(self, *args, **kwargs):
```
Could you also define some comparison operators as well? It might be useful when this gets integrated into python3

ishanrai05 replied:
yes, I am working on it. also while switching to python3 we could use `OrderedDict` instead of `dict` in ProxyDict, we won't have to sort the dictionary for getting hash keys then.

alexanderrichards replied:
If you feel that the `OrderedDict` approach is better and should be what we are using for python 3 then once this code is merged into the python 3 branch please make the change and open a PR against the python 3 branch.

ishanrai05 replied:
Yes.. Sure @alexanderrichards

mesmith75 commented on ganga/GangaCore/GPIDev/Base/CoW.py:
```diff
@@ -0,0 +1,504 @@
+                    self.remove(old)
+                    self.add(new)
+
+        except Exception as e:
```
I'm not sure I like this treatment of the exception. Can we catch a particular one and deal with it properly?

ishanrai05 replied:
yes. on it.

mesmith75 commented:
Is this quicker than the current copying? When I did the splitter copy for jobs I ended up taking out a whole load of code that just slowed it down - it looks like this is quite dense so I wonder if it is still quick.

ishanrai05 committed:
[CoW] Code Refactoring

ishanrai05 committed:
[CoW] Adds weakref to Registry slots

ishanrai05 committed:
[CoW] Comments __hash__ in GangaList

The __hash__ function in GangaList is interfering with __hash__
function in Copy-On-Write. Need to find a permanent solution for
this.

ishanrai05 committed:
[CoW] Adds Test for Bool

ishanrai05 committed:
[CoW] Adds Test for str

ishanrai05 committed:
[CoW] Adds Test for None

ishanrai05 committed:
[CoW] Adds Test for int

ishanrai05 committed:
[CoW] Adds Test for float

ishanrai05 committed:
[CoW] Adds Test for set

ishanrai05 committed:
[CoW] Adds Test for tuple

ishanrai05 committed:
[CoW] Adds Test for list

ishanrai05 committed:
[CoW] Adds Test for dict

ishanrai05 committed:
[CoW] Finish adding tests

ishanrai05 committed:
[CoW] change to python3

ishanrai05 committed:
[CoW]: Remove proxy layers before creating hash

ishanrai05 committed:
[CoW] Change weakref to strongref before pickling

ishanrai05 committed:
[CoW] Manage hashes

ishanrai05 committed:
[CoW] Remove getstate and setstate

ishanrai05 committed:
[CoW] Use cloudpickle instead of pickle

ishanrai05 committed:
[CoW] Extend CoW to Job

ishanrai05 committed:
[CoW] Changed the protocol for cloudpickle to default protocol

ishanrai05 committed:
[CoW] Rewrote hash for GangaObjects

ishanrai05 committed:
[CoW] Define getstate for GangaObjects to allow pickling

The default protocol for pickle requeres us to define __getstate__
where we use __slots__.

ishanrai05 committed:
[CoW] Changed the default value for Job ID to 0.

The current default value for job id is set to an emptly literal.
This is causing list indexing error when using it with CoW.

ishanrai05 committed:
[CoW] Add dependencies in requirements.txt

egede reviewed (commented):
Do you need the specific version number here? If not, I would take it away. These locked in version often cause trouble in the future.

ishanrai05 committed:
[CoW] Fix test errors

ishanrai05 commented:
@egede The Jenkins tests are not running. 
I have removed the version number.

alexanderrichards commented:
I just killed the jenkins jobs as they were indeed stuck. I'll let the backlog clear to determine if it was just for this PR they were stuck or if there is a more general problem. It could be that the docker image store has filled up again.

ishanrai05 commented:
I can't see why jenkins tests are failing. The site is unreachable.

egede commented:
[fail.txt](https://github.com/ganga-devs/ganga/files/3791402/fail.txt)
I see a large number of failures in the automatic testing. Please see the file attached here. 

You should be able to run the testing yourself using the Dockerfile https://github.com/ganga-devs/ganga/blob/develop/ganga/GangaCore/test/Dockerfile 

Just build the image and run it. It should be fairly self explaining what it does.

mesmith75 commented:
What is the plan for this?
